063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,207 @@
+
+At the moment Strimzi relies on these default JMX reporters and uses [jmx_exporter](https://github.com/prometheus/jmx_exporter) which is a Java agent that retrieves metrics via JMX and exposes them over an HTTP endpoint in the Prometheus format. Then Prometheus is configured to scrape that endpoint to retrieve the Kafka metrics.
+
+![Current Situation](./images/063-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer metrics. It’s named `FilteringJmxReporter` because it extends the `JmxReporter` class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+2. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for Kafka metrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter`. Both jmx_exporter and the built-in reporter offer the flexibility to choose which metrics to collect.
+3. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. This prevents using metrics like an API, whether this is to build Grafana dashboards or for the operator to rely on metrics.
+4. Due to the complex rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions.
+
+
+## Proposal
+
+The proposal is to build metrics reporters that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+Reporters will expose the following configurations:
+- `prometheus.metrics.reporter.listener`: The listener to expose the metrics in the format `http://<HOST>:<PORT>`, or set to `disabled` to not expose an endpoint. If the `<HOST>` part if empty the listener binds to the default interface, if it is set to `0.0.0.0`, the listener binds to all interfaces. If the `<PORT>` part is set to `0`, a random port is picked. Default: `http://:8080`.
```

Comment:
it looks weird to me that this configuration, looking for an hostname:port value, could be "disabled". Also what's the point of having a reporter which doesn't expose the metrics via a listener? Should we make it mandatory so not allowing anything which is not hostname:port and avoiding anything else like "disabled"?

Reply:
Maybe I see why you need it, because of the bridge use case? Let's continue the discussion here or on the comment I made for the bridge, as you prefer :-)

Reply:
Yes it's because of the bridge.

Reply:
And isn't it enough to not set  `prometheus.metrics.reporter.listener` parameter instead of setting it to "disabled"?

Reply:
I think it's important to have a good default value as the goal is to be usable outside of Strimzi too. Also I expect that disabling the endpoint is to be very marginal use case (ie not apply to most users), so it's best to require to explicitly set it to `disabled`.

Reply:
It is not a blocker for me and I can easily live with it as it is -> but I find this a bit strange pattern as well. Especially since you seem to accept values following some pattern (`http://<host>:<port>`) but then a completely different pattern (value `disabled`) is valid as well. What if there is a separate boolean option to enable the listener (e.g. prometheus.metrics.reporter.listener.enable`) that defaults to true? That way, you can have a good default but also perhaps an easier to work with way for disabling it?

Reply:
Yeah I agree with Jakub. Tbh this change of pattern for the possible values is weird to me. I am for the additional parameter as well for disabling it (which is false by default when not set).

Reply:
I'm fine with adding a separate configuration to disable the endpoint if you prefer. In the end you'll still need to set one configuration to disable it.

Reply:
Only thing, it's kind of a "blocker" for me :-P

Reply:
Thanks Mickael!